uint8 histogram equalization

grlee77 opened the pull request:
## Description

This PR adds a histogram equalization specific to uint8 images (as discussed in #5989). This is consistent with OpenCV's `equalizeHist`.

This equalization is faster because it uses a simple lookup table rather than relying on floating point interpolation of values. A new `method` kwarg was added to allow selection between the 'float' and 'uint8' implementations. By default the existing floating point interpolation is used for backwards compatibility.

For our `camera` image, the existing floating point equalization takes 6.22 ms on my system while the uint8 version proposed here takes 0.90 ms. For reference, OpenCV's `cv2.equalizeHist` takes 0.2 ms

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

grlee77 committed:
Add faster uint8 implementation of equalize_hist

Avoids use of np.interp via a simple lookup table. The uint8 implementation is only used
if the user sets preserve_uint8_dtype=True.

grlee77 committed:
change from boolean preserve_uint8_dtype flag to method kwarg

grlee77 committed:
test invalid inputs

isort imports

pep8

grlee77 committed:
update test style, removing unneeded testing import

grlee77 committed:
make method keyword-only

rfezzani reviewed (commented):
Thank you @grlee77, the performance gain is impressive :rocket:.
Can't we allow this method for dtype other then uint8? I mean that if the input image is not uint8 we can convert it using `img_as_ubyte`...
If you agree with this suggestion I suggest to replace the `method` argument by a boolean argument `use_lut`. What do you think?

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -368,14 +414,24 @@ def equalize_hist(image, nbins=256, mask=None):
+    elif method != 'float':
+        raise ValueError("method must be 'uint8' or 'float'")
     if mask is not None:
         mask = np.array(mask, dtype=bool)
```
This conversion is required before the method selection (`uint8` or `float`)

grlee77 replied:
right, nice catch

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -346,7 +345,49 @@ def cumulative_distribution(image, nbins=256):
+
+    Returns
+    -------
+    out : float array
```
```suggestion
    out : uint8 ndarray
```

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -346,7 +345,49 @@ def cumulative_distribution(image, nbins=256):
+
+    Parameters
+    ----------
+    image : array
```
```suggestion
    image : uint8 ndarray
```

grlee77 commented:
> Can't we allow this method for dtype other then uint8? I mean that if the input image is not uint8 we can convert it using img_as_ubyte... If you agree with this suggestion I suggest to replace the method argument by a boolean argument use_lut. What do you think?

I'm not sure about allowing conversion to ubyte when the flag is set. Certainly, the method could be adapted for different upper limits than 255 in the LUT.

grlee77 committed:
Apply suggestions from code review

Co-authored-by: Riadh Fezzani <rfezzani@gmail.com>

grlee77 committed:
mask asarray call needs to be earlier

grlee77 committed:
update equalize_hist_uint8->equalize_hist_uint

now adapted to work for other unsigned integer ranges

grlee77 committed:
fix typo

rfezzani commented:
@grlee77,
> I'm not sure about allowing conversion to ubyte when the flag is set.

We can add a warning when a conversion is made specifying that this warning can be avoided by not using this method or by converting the input to ubyte. Something like `The input image will be converted to ubyte: the "uint8" method needs the input image to be unsigned int. To avoid this warning you may use the "float" method or convert the input image to unsigned int.`

> Certainly, the method could be adapted for different upper limits than 255 in the LUT.

Exactly, you already defined the `lut_max` argument in `_equalize_hist_uint`. BTW, `lut_max` default value may be `None` and be set to `np.iinfo(image.dtype).max` internally when not specified :wink:.

> Also, not sure that use_lut will be a friendly argument name for newbies (may not be obvious that lut=lookup table)

We can specify this in the documentation or find a better argument name :slightly_smiling_face:

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -346,7 +345,56 @@ def cumulative_distribution(image, nbins=256):
+    cdf = hist.cumsum()
+
+    # make sure first bin corresponds to 0 in the lookup table
+    first_nonzero = np.nonzero(cdf)[0][0]
+    cdf[first_nonzero:] = cdf[first_nonzero:] - cdf[first_nonzero]
```
```suggestion
    first_nonzero = np.flatnonzero(cdf)[0]
    cdf[first_nonzero:] -= cdf[first_nonzero]
```

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -346,7 +345,56 @@ def cumulative_distribution(image, nbins=256):
 
 
-def equalize_hist(image, nbins=256, mask=None):
+def _equalize_hist_uint(image, mask=None, lut_max=255):
```
`lut_max` is not documented :slightly_smiling_face:

grlee77 replied:
now fixed (forgot to come back to this PR!)

grlee77 committed:
Update skimage/exposure/exposure.py

Co-authored-by: Riadh Fezzani <rfezzani@gmail.com>

grlee77 committed:
Merge remote-tracking branch 'upstream/main' into equalize_hist-uint8

grlee77 committed:
DOC: describe lut_max in _equalize_hist_uint docstring

grlee77 committed:
Update skimage.exposure benchmark suite

Use two images of different size: moon is (512, 512), eagle is (2019, 1826)

Parameterize across dtypes: some functions have a uint8-specific fast path

grlee77 committed:
move rgb2gray, rgba2gray imports back within is_low_contrast

grlee77 commented:
```
benchmark_exposure.ExposureSuite.time_equalize_hist                                                                                                                    [ 70.00%] ··· ============ ============ ============ ============
              --                           dtype                 
              ------------ --------------------------------------
               image_name     uint8       float32      float64   
              ============ ============ ============ ============
                  moon       908±7μs     6.13±0.1ms   5.85±0.1ms 
                 eagle      15.3±0.3ms   108±0.7ms     107±1ms   
              ============ ============ ============ ============
```

rfezzani commented on skimage/exposure/exposure.py:
```diff
@@ -825,7 +891,7 @@ def is_low_contrast(image, fraction_threshold=0.05, lower_percentile=1,
 
     if image.ndim == 3:
-        from ..color import rgb2gray, rgba2rgb  # avoid circular import
+        from ..color.colorconv import rgb2gray, rgba2rgb  # no circular import
```
I tested it locally, it seems that this circular import disappeared :tada: Can you confirm @grlee77?

grlee77 replied:
Did you test with EAGER_IMPORT=1 defined? (It failed on CI with a circula import on a previous run, so I  added it back in)

rfezzani replied:
> Did you test with EAGER_IMPORT=1 defined? (It failed on CI with a circula import on a previous run, so I added it back in)

Nop, I just ran `pytest` locally :sweat_smile:... Too bad :confused:

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
